modules/javafx.graphics/src/main/java/com/sun/javafx/util/Utils.java

```diff
@@ -143,6 +147,353 @@ public static double nearest(double less, double value, double more) {
+     */
+    public static double interpolateDiscrete(double from, double to, double t) {
+        return t < 0.5 ? from : to;
+    }
+
+    /**
+     * Computes an intermediate list that consists of the pairwise interpolation between two lists,
+     * using the following rules:
+     * <ol>
+     *     <li>The size of the returned list corresponds to the size of the second list.
+     *     <li>If the first list has fewer elements than the second list, the missing elements are copied
+     *         from the second list.
+     *     <li>If the first list has more elements than the second list, the excess elements are discarded.
+     *     <li>If the intermediate list is shallow-equal to the first list passed into the method (i.e. its
+     *         elements are references to the same objects), the existing list is returned.
+     *     <li>If a new list is returned, it is unmodifiable.
+     * </ol>
+     *
+     * @param firstList the first list, not {@code null}
+     * @param secondList the second list, not {@code null}
+     * @return the intermediate list
+     */
+    public static <T extends Interpolatable<T>> List<T> interpolateListsPairwise(
+            List<T> firstList, List<T> secondList, double t) {
+        Objects.requireNonNull(firstList, "firstList");
+        Objects.requireNonNull(secondList, "secondList");
+
+        if (secondList.isEmpty()) {
+            return firstList.isEmpty() ? firstList : secondList;
+        }
+
+        int listSize = firstList.size();
+
+        // For small equisized lists (up to 8 elements), we use an optimization to prevent the allocation
+        // of a new array in case the intermediate list would be equal to the existing list.
+        if (listSize <= 8 && listSize == secondList.size()) {
+            return interpolateEquisizedListsPairwise(firstList, secondList, t);
+        }
+
+        @SuppressWarnings("unchecked")
+        T[] newArray = (T[])new Interpolatable[secondList.size()];
+        boolean equal = firstList.size() == secondList.size();
+
+        for (int i = 0, firstListSize = firstList.size(); i < newArray.length; ++i) {
+            if (firstListSize > i) {
+                newArray[i] = firstList.get(i).interpolate(secondList.get(i), t);
+                equal &= newArray[i] == firstList.get(i);
+            } else {
+                newArray[i] = secondList.get(i);
+            }
+        }
+
+        return equal ? firstList : new UnmodifiableArrayList<>(newArray, newArray.length);
+    }
+
+    /**
+     * Computes an intermediate list that consists of the pairwise interpolation between two lists
+     * of equal size, each containing up to 8 elements.
+     * <p>
+     * This method is an optimization: it does not allocate memory when the intermediate list is
+     * shallow-equal to the list that is passed into this method, i.e. its elements are references
+     * to the same objects. The existing list is returned in this case.
+     */
```

Comment:
There's no performance problem in the sense that my test application runs slowly in any noticeble way. It's one more piece of trying to minimize the allocation of objects in the very common case where nothing has changed.

Do you think it is preferrable to appeal to JIT magic instead of explicitly coding the behavior? With your other suggestion of using `switch`, the code looks much cleaner already.

Reply:
> Do you think it is preferrable to appeal to JIT magic instead of explicitly coding the behavior?

If the code leads to less questions and is more readable for it, and there is no reason to optimize in the first place (ie. you didn't see a problem), then yes, definitely.

Premature optimization only leads to questions by reviewers and future maintainers (they will wonder why the code was optimized, but probably assume there was a good reason for it).  Optimized code is harder to reason about and harder to refactor.  For example, reading your code I've been left wondering:

- Why is it significant to return a specific instance?  Are you using `==` to compare these in callers?  That's very counter to the nature of Java, and so deserves at a minimum justification ("performance") and warning ("careful, callers use `==`")
- The extra code paths for `RandomAccess` or not, in an attempt to avoid an iterator allocation (which may not even happen when inlined).  Is this even happening in practice (are there any non-`RandomAccesss` lists used?)  Is the performance gain worth the extra complexity?  Is it worth not having these functions inline into callers, as larger functions mean JIT won't inline them?
- Avoiding allocating small arrays; was this a performance problem that would justify writing a function that is now too large to inline? You trading potential JIT inlining for an almost free heap allocation (Java allocation's are the equivalent of a pointer increment + background GC activity on likely **idling** CPU cores).

Would it not have been much better to write the code as simple as possible first, get it passed review well understood by everyone, then maybe, possibly get back to it later with an optimizing PR if there is a performance problem?  That last bit is entirely optional and may never happen if it runs out these few transition allocations are but a drop in the bucket of the overall CSS performance.

Reply:
Sorry if this came across as a bit negative, it's not intended in that way.  I just think optimizing too early has far more drawbacks than it has benefits. In part this is also a reflection on a lot of FX code; some code is clearly optimized from a C/C++ developer perspective which not only translates poorly to Java, but also makes the code "too dangerous" to modify.

Reply:
I don't understand this argument at all. Why would it be reasonable to _only_ build a more efficient architecture _if_ there's such a huge problem with the "naive" alternative that this alone would be readily apparent? If taken to the extreme, this line of thinking leads to death by a thousand papercuts, no single issue big enough that it's a problem all by itself, but taken together it amounts to sluggish performance.

Here's the thing: this code could reasonably be executed tens of thousands of times per second. Think of a complex user interface that changes its visual state for many nodes at once. What you seem to perceive as a micro-optimization is actually a part of an architecture that involves many types in the `Background` and `Border` area.

The basic idea is: if I can rely on `interpolate()` returning existing instances (either the start value or the end value) if nothing has changed, then the "parent object's" interpolate method can use an identity comparison (which is basically for free) to choose whether to return itself (or the end value), or return a new instance.

Let's look at the `Border.interpolate()` implementation:
```java
@Override
public Border interpolate(Border endValue, double t) {
    Objects.requireNonNull(endValue, "endValue cannot be null");
    if (t <= 0) return this;
    if (t >= 1) return endValue;

    // interpolateListsPairwise() is implemented such that it returns existing list instances
    // (i.e. the 'this.images' or 'endValue.images' arguments) as an indication that the result
    // is shallow-equal to either of the input arguments. This allows us to very quickly detect
    // if we can return 'this' or 'endValue' without allocating a new Border instance.
    List<BorderImage> newImages = images == endValue.images ?
        images : InterpolationUtils.interpolateListsPairwise(images, endValue.images, t);

    List<BorderStroke> newStrokes = strokes == endValue.strokes ?
        strokes : InterpolationUtils.interpolateListsPairwise(strokes, endValue.strokes, t);

    if (images == newImages && strokes == newStrokes) {
        return this;
    }

    if (endValue.images == newImages && endValue.strokes == newStrokes) {
        return endValue;
    }

    return new Border(newStrokes, newImages);
}
```

`interpolateListsPairwise()` _already knows_ whether both lists are equal, which makes the `return this` and `return endValue` branches after the interpolation basically free. The alternative to this would be to either pointlessly call `equals()` on a large object graph again and again, or to always return a new `Border` instance.

This optimization scheme works precisely because all types that are involved make sure to return existing instances when the result is known to be interchangable with an existing instance. This is also the reason why your other suggestion of returning `List.of()` instead of an existing empty list wouldn't work.

Reply:
I think you said that you didn't actually test this, so we don't know if this will perform better. It's based on an assumption of what Java will do, and on assumptions what is "bad" and what is "good".  Without having testing this, you are optimizing blind.  I would be entirely satisfied if you had said "I tested this with a few thousand transitions, the allocation turned out to be a hot spot, so I optimized it which is the reason this code looks a bit out of the ordinary".

I also would have liked a comment explaining any pitfalls introduced by the optimization; the use of `==` by callers is unusual, and a crucial piece of information when I see code like:

        if (secondList.isEmpty()) {
            return firstList.isEmpty() ? firstList : secondList;
        }

It would deserve at least an explanation that you're specifically returning one of the passed in arguments so callers can do reference equality checks.

So you may have been wasting your time, and perhaps future maintainer time as well, making code less flexible, and harder to understand, maybe containing new bugs and edge cases for no real reason.

Java's JIT is very finicky.  For example, an earlier comment asked if the first `if` case is the most common one.  The assumption being that this will improve performance.  However, you can't make that assumption at all (I've done tests that show it's almost impossible to predict which "if order" will be faster on the JVM); re-arranging a function or "unrolling" it can have large negative consequences if that means the JIT will no longer inline your function.

Now, I will give you that this **probably** will be faster, but I'm making an assumption here... I really can't be sure.  That still leaves some explanatory comments that I think you should add with the assumptions that were made for this optimization.

>  If taken to the extreme, this line of thinking leads to death by a thousand papercuts, no single issue big enough that it's a problem all by itself, but taken together it amounts to sluggish performance.

No, when hunting for performance problems, it will be readily apparent what total process is taking up all the time, and how often parts of it are being called and how much time they take in total.  Put a profiler on it, and see where the actual time is being spent on a case with a lot of transitions if you want to optimize it.

What optimization generally does is it that it makes code harder to understand (almost unavoidable when optimizing), the code will make more assumptions (which is where performance is gained), and will often require more debugging time and dealing with edge cases (more complex code has more bugs and edge cases, requiring more tests).  

When applied randomly to a code base (ie. without profiling) it only makes things harder for no tangible benefits, and closes off easy improvements later (undoing optimizations for a better alternative has been the topic of several of my PR's).  So if you're going to make something more complicated, and harder to refactor and extend later (which can have far greater pay-offs), then I'm saying there should be a good reason for it.

Reply:
I think what is being done in this PR is fine, even though your arguments @hjohn are perfectly valid.  One feature that I think has value is the fact that this design tries to limit new allocations, and I think it's a good thing.

One interesting point you bring up is that we have no good performance test for CSS.  What would represent a good performance test?  Create a thousand of labels and measure the time it takes to update?  Does it matter if it's a label or something else?  And do we have a reliable way to measure?  What do you think?